hikari/impl/bot.py

```diff
@@ -23,589 +23,481 @@
-            self._event_manager = stateless_event_manager.StatelessEventManagerImpl(app=self, intents=intents)
-            _LOGGER.info("this application is stateless, cache-based operations will not be available")
-        else:
-            self._cache = cache_impl.StatefulCacheImpl(app=self, intents=intents)
-            self._guild_chunker = guild_chunker_impl.StatefulGuildChunkerImpl(app=self)
-            self._event_manager = stateful_event_manager.StatefulEventManagerImpl(
-                app=self, cache=self._cache, intents=intents
-            )
 
-        self._voice = voice.VoiceComponentImpl(self, self._event_manager)
-
-    def __del__(self) -> None:
-        # If something goes wrong while initializing the bot, `_start_count` might not be there.
-        if hasattr(self, "_start_count") and self._start_count == 0:
-            # TODO: may remove this as it causes issues with tests. Provisionally implemented only.
-            warnings.warn(
-                "Looks like your bot never started. Make sure you called bot.run() after you set the bot object up.",
-                category=errors.HikariWarning,
-            )
+        # We populate these on startup instead, as we need to possibly make some
+        # HTTP requests to determine what to put in this mapping.
+        self._shards: typing.Dict[int, shard.GatewayShard] = {}
 
     @property
     def cache(self) -> cache_.Cache:
-        # <<inherited docstring from traits.CacheAware>>
         return self._cache
 
     @property
-    def chunker(self) -> guild_chunker_.GuildChunker:
-        # <<inherited docstring from traits.ChunkerAware>>
-        return self._guild_chunker
+    def chunker(self) -> chunker_.GuildChunker:
+        return self._chunker
 
     @property
     def dispatcher(self) -> event_dispatcher.EventDispatcher:
-        # <<inherited docstring from traits.DispatcherAware>>
-        return self._event_manager
+        return self._events
 
     @property
-    def entity_factory(self) -> entity_factory_impl.EntityFactoryImpl:
-        # <<inherited docstring from traits.EntityFactoryAware>>
+    def entity_factory(self) -> entity_factory_.EntityFactory:
         return self._entity_factory
 
     @property
-    def event_factory(self) -> event_factory_impl.EventFactoryImpl:
-        # <<inherited docstring from traits.EventFactoryAware>>
+    def event_factory(self) -> event_factory_.EventFactory:
         return self._event_factory
 
     @property
     def executor(self) -> typing.Optional[concurrent.futures.Executor]:
-        # <<inherited docstring from traits.ExecutorAware>>
         return self._executor
 
     @property
     def heartbeat_latencies(self) -> typing.Mapping[int, float]:
-        # <<inherited docstring from traits.ShardAware>>
-        return {shard_id: shard.heartbeat_latency for shard_id, shard in self._shards.items()}
+        return {s.id: s.heartbeat_latency for s in self._shards.values()}
```

Comment:
Should we do `self._shards.copy()` to avoid an edge case error caused by someone calling a method or property that iterates through self._shards while the bot is rolling up and creating shards objects?

Reply:
this would only occur if they did it from another thread, and the application isn't marked as being threadsafe, so I am not sure.

Threadsafety should probably go in a separate issue as I doubt this is the only place this may cause a headache later.